sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -683,9 +685,28 @@ abstract class HadoopFsRelation private[sql](
       }
     }
 
-    buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf)
+    if (bucketSpec.isEmpty || !sqlContext.conf.bucketingEnabled()) {
+      buildInternalScan(requiredColumns, filters, allInputFiles, broadcastedConf)
+    } else {
+      // For each bucket id, firstly we get all files belong to this bucket, by detecting bucket id
+      // from file name. Then read these files into a RDD(use one-partition empty RDD for empty
+      // bucket), and coalesce it to one partition. Finally union all bucket RDDs to final result.
+      val perBucketRows = (0 until bucketSpec.get.numBuckets).map { bucketId =>
+        val inputStatuses = allInputFiles.filter(f => getBucketId(f) == bucketId)
```

Comment:
Is getBucketId cheap? Might be nice to just map and collect all the ids instead of making this O(num buckets * num files)